Swik 1280 username case insensitive

TBoonX opened the pull request:
I made that usernames have to be alphanumeric and could also contain -._~
I added checking and parsing. I made sure that the upper cases of usernames last.
I also improved the reading/parsing/usage of the email parameter. (**urgent**)
There is now a migration script which I described to ben and which will also be added to the migration repo.

TBoonX committed:
[SWIK-unknown_username_case_insensitive] First changes for the ticket - MongoDB Queries are wrong atm

TBoonX committed:
[SWIK-1280_username_case_insensitive] merged from  master

TBoonX committed:
[SWIK-1280_username_case_insensitive] Corrected search of email and username case insensitive

TBoonX committed:
[SWIK-1280_username_case_insensitive] Updated tests. Fixed tests database name selection. Fixed issue in util.js

TBoonX committed:
[SWIK-1280_username_case_insensitive] Now reading username from payload does trimming

TBoonX committed:
[SWIK-1280_username_case_insensitive] Created the migration script - works but have to be closed manually

TBoonX committed:
[SWIK-1280_username_case_insensitive] MAde migration script exiting itself.

TBoonX committed:
[SWIK-1280_username_case_insensitive] Merge from master

TBoonX committed:
[SWIK-1280_username_case_insensitive] Fixed missing stuff after merge

TBoonX committed:
[SWIK-1280_username_case_insensitive] Used regex for username on all routes

TBoonX committed:
[SWIK-1280_username_case_insensitive] Worked on migration code

TBoonX committed:
[SWIK-1280_username_case_insensitive] Merge code seems to work now

rmeissn commented:
Can you please resolve the conflicts of this PR? I will review it today.

TBoonX committed:
[SWIK-1280_username_case_insensitive] Merged from master

TBoonX committed:
[SWIK-1280_username_case_insensitive] Merge from master

rmeissn commented on application/controllers/handler.js:
```diff
@@ -80,7 +80,14 @@ module.exports = {
-      email: new RegExp(decodeURI(req.payload.email).replace(/\s/g,''), 'i'),
+      $and: [
+        {
+          email: new RegExp(req.payload.email, 'i')
```
use .tolowercase() instead and make sure every email in the DB is lowercase (by migrating them once and saving new emails as lowercase)

rmeissn commented on application/controllers/handler.js:
```diff
@@ -274,7 +281,7 @@ module.exports = {
           $set: {
             email:       email.toLowerCase(),
-            username:    util.parseAPIParameter(req.payload.username),
+            username:    util.parseAPIParameter(req.payload.username).replace(/\s/g,''),
```
Redundant to remove whitspaces because they are already handled at the route itself

rmeissn commented on application/controllers/handler.js:
```diff
@@ -345,15 +352,27 @@ module.exports = {
 
   getPublicUser: (req, res) => {
-    let identifier = decodeURI(req.params.identifier);
+    let identifier = decodeURI(req.params.identifier).replace(/\s/g,'');
```
Handle all chars correctly

TBoonX replied:
As the identifier is an integer or a string I validate with the correct regex if integer validations fails

rmeissn commented on application/controllers/handler.js:
```diff
@@ -385,10 +404,29 @@ module.exports = {
-      username: username
+      $and: [
+        {
+          username: new RegExp(username, 'i')
```
Partly same as first comment, make regex exact match

rmeissn commented on application/controllers/handler.js:
```diff
@@ -485,10 +529,17 @@ module.exports = {
-      email: email
+      $and: [
+        {
+          email: new RegExp(email, 'i')
```
Same as first comment

rmeissn commented on application/controllers/handler.js:
```diff
@@ -591,7 +642,14 @@ module.exports = {
-          email: new RegExp(data.email, 'i')
+          $and: [
+            {
+              email: new RegExp(data.email, 'i')
```
Same as first comment

rmeissn commented on application/controllers/handler.js:
```diff
@@ -974,7 +1033,14 @@ function isUsernameAlreadyTaken(username) {
-      email: new RegExp(email, 'i')
+      $and: [
+        {
+          email: new RegExp(email, 'i')
```
Same as first comment

rmeissn commented on application/controllers/util.js:
```diff
@@ -31,10 +31,24 @@ module.exports = {
-            username: username
+            $and: [
+              {
+                username: new RegExp(username, 'i')
```
Look for both statements if they can be reduced by the same method as in handler.js

rmeissn commented on application/controllers/util.js:
```diff
@@ -45,11 +59,11 @@ module.exports = {
           if (array.length > 0) {
             const isEMailAssigned = !(array.reduce((prev, curr) => {
-              const sameEMail = curr.email === email;
+              const sameEMail = curr.email.toLowerCase() === email.toLowerCase();
```
If in database always lowercase, remove the call here

TBoonX committed:
[SWIK-1280_username_case_insensitive] Improved code as requested by roy.

TBoonX committed:
[SWIK-1280_username_case_insensitive] migration code now makes emails to lowercase. Also added small data error handling in migration code

rmeissn commented:
I've tested the Service-API:
* Using /user/{id}/profile PUT to change user information, I am able to exchange case-sensitive chars of the username. E.g. change rmeissn to RmeiSsn
* I can't test /usergroup/createorupdate in Swagger as members is never issued as an array (maybe it's possible to fix it)

I haven't tested social login and usergroups, but code looks good.

TBoonX committed:
[SWIK-1280_username_case_insensitive] Reading a users public profile now searches case insensitive

TBoonX committed:
[SWIK-1280_username_case_insensitive] Switched swagger input type from form to json for route /usergroup/createorupdate

TBoonX commented:
@rmeissn 
* We liked to keep the users preference thus I would say its okay to change the case sensitivity of the chars. It is not used in the frontend thus I would like to keep this logic.
* I fixed the swagger problem

Additionally I added case insensitivity to the route which returns the users public profile, as requested by kostis: https://slidewiki.atlassian.net/wiki/pages/viewpage.action?pageId=43251468

TBoonX committed:
[SWIK-1280_username_case_insensitive] Fixed route getpublicuser

rmeissn commented:
About the Migration script: I think you can simplify it a lot.
What you want to do is: map users (email to lowercase and remove not allowed chars from usernames) and group users by username (to group users with the same usernames). Subsequently map all users to users and map groups of users to users (username is modified, e.g. added "_", emit() is called several times for a group).
Thus you would be able to get rid of all the reduce functions (and thus LOC and logic).

TBoonX commented:
OK this could be done later. Thus @kadevgraaf please merge it.
